### PR TITLE
Subtract: Remove the intersection of two ranges

### PR DESCRIPTION
**Pull request by bramski**
Subtracting two ranges will result in an array of ranges:
empty set -- the second range is equal or greater than the first
single set -- the second set overlaps the beginning or end of the first or does not intersect at all
two ranges -- the second set is within the first and slices it into two new pieces

### CODE REVIEW

File: datetimerange/__init__.py [hunk 1]
```diff
@@ -635,6 +635,61 @@ def intersection(self, x):
...
+            .. code:: python
+
+            from datetimerange import DateTimeRange
+            dtr0 = DateTimeRange("2015-03-22T10:00:00+0900", "2015-03-22T10:10:00+0900")
+            dtr1 = DateTimeRange("2015-03-22T10:05:00+0900", "2015-03-22T10:15:00+0900")
+            dtr0.subtract(dtr1)
+        :Output:
+          .. parsed-literal::
+
+                [2015-03-22T10:00:00+0900 - 2015-03-22T05:00+0900]
```
Comment (mattasaminew) [inline datetimerange/__init__.py:660]: This should be `[2015-03-22T10:00:00+0900 - 2015-03-22T10:05:00+0900]`

File: test/test_dtr.py [hunk 2]
```diff
@@ -730,6 +730,82 @@ def test_exception(self, lhs, rhs, expected):
             lhs.is_intersection(rhs)
 
 
+class TestDateTimeRange_subtract:
+    @pytest.mark.parametrize(
+        ["lhs", "rhs", "expected"],
+        [
```
Comment (mattasaminew) [inline test/test_dtr.py:736]: Maybe add a test case for when the `lhs` and `rhs` do not overlap?

File: datetimerange/__init__.py [hunk 3]
```diff
@@ -635,6 +635,61 @@ def intersection(self, x):
...
+            .. code:: python
+
+            from datetimerange import DateTimeRange
+            dtr0 = DateTimeRange("2015-03-22T10:00:00+0900", "2015-03-22T10:10:00+0900")
+            dtr1 = DateTimeRange("2015-03-22T10:05:00+0900", "2015-03-22T10:15:00+0900")
+            dtr0.subtract(dtr1)
+        :Output:
+          .. parsed-literal::
+
+                [2015-03-22T10:00:00+0900 - 2015-03-22T10:05:00+0900]
```
Comment (bramski) [inline datetimerange/__init__.py:660]: @mattasaminew addressed

File: test/test_dtr.py (hunk 2 above)
Comment (bramski) [reply test/test_dtr.py:736]: Addressed. See third test case.

**Review by mattasaminew — APPROVED** [review]
Looks good to me

File: datetimerange/__init__.py [hunk 4]
```diff
@@ -635,6 +635,61 @@ def intersection(self, x):
             end_time_format=self.end_time_format,
         )
 
+    def subtract(self, x):
+        """
+        Remove a time range from this one and return the result.
+        The result will be [self.copy()] if the second range does not overlap the first
+        The result will be [] if the second range wholly encompasses the first range
+        The result will be [new_range] if the second range overlaps one end of the range
+        The result will be [new_range1, new_range2] if the second range is an internal sub range of the first
```
Comment (thombashi) [inline datetimerange/__init__.py:644]: The above lines are a little bit hard to read in the HTML after the docs build.
How about modifying like the followings:

```
        Remove a time range from this one and return the result.

        - The result will be ``[self.copy()]`` if the second range does not overlap the first
        - The result will be ``[]`` if the second range wholly encompasses the first range
        - The result will be ``[new_range]`` if the second range overlaps one end of the range
        - The result will be ``[new_range1, new_range2]`` if the second range is an internal sub range of the first
```

**Review by thombashi — CHANGES_REQUESTED** [review]

**Comment by thombashi** [comment]
Please apply code formatter. You can do that as follows:

```
make setup
make fmt
```

File: datetimerange/__init__.py [hunk 5]
```diff
@@ -635,6 +635,61 @@ def intersection(self, x):
...
+            return [DateTimeRange(start_datetime=self.start_datetime, end_datetime=overlap.start_datetime,
+                                  start_time_format=self.start_time_format, end_time_format=self.end_time_format)]
+
+        # Case 5, underlap, two new ranges are needed.
+        return [
+            DateTimeRange(start_datetime=self.start_datetime, end_datetime=overlap.start_datetime,
+                          start_time_format=self.start_time_format, end_time_format=self.end_time_format),
+            DateTimeRange(start_datetime=overlap.end_datetime, end_datetime=self.end_datetime,
+                          start_time_format=self.start_time_format, end_time_format=self.end_time_format)
+        ]
```
[the remark this answers is not in the document]
Comment (bramski) [reply datetimerange/__init__.py:691]: Ah I see. It's a zero length range. I would say yeah it's not really a valid case for breaking up the range and it should not cut it into two.

File: docs/pages/examples/Make_a_subtracted_time_range.rst [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+:Sample Code:
+    .. code:: python
+
+        from datetimerange import DateTimeRange
+        time_range = DateTimeRange("2015-01-22T09:50:00+0900", "2015-01-22T10:00:00+0900")
+        x = DateTimeRange("2015-01-22T09:55:00+0900", "2015-01-22T09:56:00+0900")
+        time_range.subtract(x)
+
+:Output:
+    ::
```
[the remark this answers is not in the document]
Comment (bramski) [reply docs/pages/examples/Make_a_subtracted_time_range.rst:12]: Addressed.

File: test/test_dtr.py [hunk 7]
```diff
@@ -730,6 +730,94 @@ def test_exception(self, lhs, rhs, expected):
...
+            ],
+            [
+                DateTimeRange("2015-01-22T09:50:00+0900", "2015-01-22T10:30:00+0900"),
+                DateTimeRange("2015-01-22T10:40:00+0900", "2015-01-22T10:50:00+0900"),
+                [DateTimeRange("2015-01-22T09:50:00+0900", "2015-01-22T10:30:00+0900")],
+            ],
+            [
+                DateTimeRange("2015-01-22T09:50:00+0900", "2015-01-22T10:30:00+0900"),
+                DateTimeRange("2015-01-22T09:55:00+0900", "2015-01-22T09:55:00+0900"),
+                [DateTimeRange("2015-01-22T09:50:00+0900", "2015-01-22T10:30:00+0900")],
```
Comment (bramski) [inline test/test_dtr.py:755]: @thombashi test case to cover that case.

File: datetimerange/__init__.py [hunk 8]
```diff
@@ -635,6 +635,88 @@ def intersection(self, x):
...
+                dtr1 = DateTimeRange("2015-03-22T10:05:00+0900", "2015-03-22T10:15:00+0900")
+                dtr0.subtract(dtr1)
+        :Output:
+          .. parsed-literal::
+
+                [2015-03-22T10:00:00+0900 - 2015-03-22T10:05:00+0900]
+        """
+        overlap = self.intersection(x)
+        # No intersection, return a copy of the original
+        if not overlap.is_set() or overlap.get_timedelta_second() <= 0:
```
Comment (bramski) [inline datetimerange/__init__.py:665]: @thombashi covers the case where you attempt to subtract a zero length range.

**Comment by bramski** [comment]
@thombashi looks like the test failures are due to python 3.5. This is all passing locally.

**Review by thombashi — APPROVED** [review]
LGTM.
Thank you for your pull request.

**Comment by thombashi** [comment]
> @thombashi looks like the test failures are due to python 3.5. This is all passing locally.

I'll fix CI.

**Pull request merged**